src/librustc_mir/transform/const_prop.rs

```diff
@@ -771,8 +783,9 @@ impl CanConstProp {
             // FIXME(oli-obk): lint variables until they are used in a condition
             // FIXME(oli-obk): lint if return value is constant
             let local_kind = body.local_kind(local);
+            cpv.local_kinds.insert(local, clone_local_kind(&local_kind));
 
-            if local_kind == LocalKind::Arg || local_kind == LocalKind::Var {
+            if local_kind == LocalKind::Arg {
```

Comment:
I believe the reason we had this was so that if you had code like this:

```rust
let x = 0u32;
let y = 1u32;

if x < y {
    y - x
} else {
    x - y
}
```

we wouldn't trigger a lint on `x - y` because it wouldn't actually be hit at runtime. Stopping full const prop on user variables was crude but effective at preventing this. I think by removing this check, we'll regress this.

Reply:
I think this is definitely the way forward for const-prop, though. 

Should we open an issue about this, and let it regress* in this PR? 
Or do you think I could maybe manage to implement logic that avoids this behavior while still having const prop on user variables? :3 

* I guess we still need to see the benchmarks on this before making a call, but I had to ask!

Reply:
We can definitely run some benchmarks to see what kind of improvements we can hope for but I don't think we can merge this with the regressed behavior as that would cause invalid warnings to be issued on valid code. 

There are probably some things we could do to allow some const propagation on user variables. The simplest thing that may work would be to only propagate within the same block that the value is assigned. I believe that would be sufficient to get your test case to work.

Reply:
> invalid warnings to be issued on valid code.

AHH I understand now. That snippet gives a warning because it sees that you can try to compute `0 - 1` on `usize` types.

> The simplest thing that may work would be to only propagate within the same block that the value is assigned. I believe that would be sufficient to get your test case to work.

Oh no, that makes me want to implement the full thing. Dang it. However... full flow-analysis would require new machinery that isn't in place yet, right? In that case, it might be worth it to do what you say and leave the unfinished business filed as an Issue or something similar.

Reply:
We do have a dataflow analysis framework in rustc. I've been meaning to migrate this pass to use it but haven't gotten around to it yet. That's probably going to be a much larger change than just getting this test case to work. 

Personally, I'd recommend starting with the smaller change first to get a bit more familiar with rustc internals. You can always pick up the larger change after that 🙂

Reply:
That seems very sensible to me! :)

Reply:
> The simplest thing that may work would be to only propagate within the same block that the value is assigned. I believe that would be sufficient to get your test case to work.

I'm currently working on getting the function call-site arguments propagated as well, but I'd rather ask right now than later... where should this change be? I'm still working on getting my bearings around the MIR codebase, so I don't think I know where this would happen... yet :thinking:

Reply:
So what I'm thinking is something like this:

- We should have a new `ConstPropMode` which means that we can only propagate within the same `BasicBlock` that the local is assigned in.

https://github.com/rust-lang/rust/blob/4ca5fd2d7b6b1d75b6cb8f679e8523fb3e7b19e2/src/librustc_mir/transform/const_prop.rs#L768-L775

- Set that here for `LocalKind::Var`

https://github.com/rust-lang/rust/blob/4ca5fd2d7b6b1d75b6cb8f679e8523fb3e7b19e2/src/librustc_mir/transform/const_prop.rs#L798

- I think we don't need this change then (but I'm not sure I completely understand why it's required currently in the PR).

https://github.com/rust-lang/rust/pull/71298/files#diff-9e103702275cbef342c2decd3395bf3bR818-R822

- Add a field to track what locals we need to forget at the end of a block here. (I recommend using `BitSet<Local>`)

https://github.com/rust-lang/rust/blob/4ca5fd2d7b6b1d75b6cb8f679e8523fb3e7b19e2/src/librustc_mir/transform/const_prop.rs#L325-L338

- The logic around here needs to be adjusted to allow the mode but, in the new mode, call `BitSet<Local>::insert()` so that we know we need to remove the const value when the block completes.

https://github.com/rust-lang/rust/blob/4ca5fd2d7b6b1d75b6cb8f679e8523fb3e7b19e2/src/librustc_mir/transform/const_prop.rs#L859-L880

- As the last step in `visit_terminator()`, loop over every local we have to forget and call `self.remove_const(local)`.

https://github.com/rust-lang/rust/blob/4ca5fd2d7b6b1d75b6cb8f679e8523fb3e7b19e2/src/librustc_mir/transform/const_prop.rs#L901

I think most of the failing tests are because we aren't forgetting the values for variables when we should be which is having wacky effects on the MIR.

Reply:
> I think we don't need this change then (but I'm not sure I completely understand why it's required currently in the PR).

I don't understand it 100% yet, but I did it as instructed by @oli-obk (it was in order to not introduce more bugs I think? Oli please correct me on this). 

I think that @eddyb also had thoughts on that part of the code, although if I remember his words correctly, it was just one of many places that required a bit of love.

Reply:
> We should have a new `ConstPropMode` which means that we can only propagate within the same `BasicBlock` that the local is assigned in.

Not restricting propagating _from_ it, right?

In that case we would have the following cases:

```rust
enum ConstPropMode {
    /// The `Local` can be propagated into and reads of this `Local` can also be propagated.
    FullConstProp,
    /// The `Local` can be propagated into and from its own block.
    ConstPropInsideOwnBlock,
    /// The `Local` can be propagated into but reads cannot be propagated.
    OnlyPropagateInto,
    /// No propagation is allowed at all.
    NoPropagation,
}
```

Is that fine? I ask because it feels a bit like we're restricting on two independent fronts: the **direction of propagation** (into and from) and the **radius of propagation** (1 block or infinite blocks). So, in a sense, it feels like we'd want to make it a 2D system (like with 2 independent enums). 

But of course that's just the shape of the type, and not its meaning. Hence me asking, in case we need to think this a bit more before I start testing implementations. I'm guessing that we don't care because the only interesting cases might be just these 4. Is that correct?